### PR TITLE
🌱 add support for topology managed fields

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR proposes a solution for ensuring the topology controller handles nicely the removal of fields from the template.

The implementation borrows the idea of managed fields from server-side apply in Kubernetes; when executing a change on templates we keep track of the fields the template expressed opinions on (provided a value).

Then, during the next reconcile on the next object, we consider all the managed fields as authoritative paths, meaning that if the corresponding value has been removed from the desired state, we enforce removal from the resulting object.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/5811

### CODE REVIEW

File: controllers/topology/internal/mergepatch/managed_paths.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// Stores the list of paths managed by the topology controller into the managed field annotation.
+// NOTE: The topology controller is only concerned about managed paths in the spec; given that
+// we are dropping spec. from the result to reduce verbosity of the generated annotation.
+// NOTE: Managed paths are relevant only for unstructured objects where it is not possible
+// to easily discover which fields have been set by templates + patches/variables at a given reconcile;
+// instead, it is not necessary to store managed paths for typed objets (e.g. Cluster, MachineDeployments)
+// given that the topology controller explicitly sets a well-known, immutable list of fields at every reconcile.
+func storeManagedPaths(obj client.Object) error {
```
Comment (sbueringer) [inline controllers/topology/internal/mergepatch/managed_paths.go:80]: Currently we ignore `ignoredPaths` when storing managedPaths. This means the following can happen:
DockerClusterTemplate
```yaml
  spec:
    template:
      metadata: {}
      spec:
        controlPlaneEndpoint:
          host: ""
          port: 0
        loadBalancer: {}
```

DockerCluster
```yaml
  kind: DockerCluster
  metadata:
    annotations:
      cluster.x-k8s.io/cloned-from-groupkind: DockerClusterTemplate.infrastructure.cluster.x-k8s.io
      cluster.x-k8s.io/cloned-from-name: quick-start-my-cluster
      topology.cluster.x-k8s.io/managed-fields: controlPlaneEndpoint.host, controlPlaneEndpoint.port,
        loadBalancer.imageRepository, loadBalancer.imageTag
```

I think we should make sure that fields which are part of ignoredPaths should not end up in the managed fields annotation. I think the same applies for fields which are **not** part of allowedPaths (essentially the managed paths should reflect the filtering we're doing when patching with allowedPaths/ignorePaths)

File: controllers/topology/internal/mergepatch/mergepatch.go [hunk 2]
```diff
@@ -124,25 +135,25 @@ type applyPathOptionsOutput struct {
...
 	// accounting for instance specific changes like we do for other maps into spec.
-	if len(in.options.authoritativePaths) > 0 {
+	if len(in.options.authoritativePaths) > 0 || len(in.options.managedPaths) > 0 {
 		authoritativePatchMap := make(map[string]interface{})
 		if err := json.Unmarshal(in.authoritativePatch, &authoritativePatchMap); err != nil {
 			return nil, errors.Wrap(err, "failed to unmarshal authoritative merge patch")
 		}
 
-		for _, path := range in.options.authoritativePaths {
+		for _, path := range append(in.options.managedPaths, in.options.authoritativePaths...) {
```
Comment (sbueringer) [inline controllers/topology/internal/mergepatch/mergepatch.go:156]: Following scenario:

Initial KCPTemplate contains:
```yaml
spec:
  template:
    spec:
      kubeadmConfigSpec:
        clusterConfiguration:
          controllerManager:
            extraArgs: { enable-hostpath-provisioner: 'true' }
```

KCP reconciled managed-field annotation contains: `kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.enable-hostpath-provisioner`

KCPTemplate is now updated to drop the entire controllerManager sub-object.

Data in mergePatch helper:
* managedPaths contains `kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.enable-hostpath-provisioner`
* authoritativePatchMap contains a (nested) key `kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs` with value `nil`

As we are only opinionated about `enable-hostpath-provisioner` and not the parent object, empty'ing the extraArgs object is not part of the patch

Comment (sbueringer) [reply controllers/topology/internal/mergepatch/mergepatch.go:156]: tl;dr I think we're not able to delete objects when we previously had opinions about nested fields.

I think I don't get the issue with DockerClusterTemplate with the loadBalancer object because in that case it's actually an empty object in the DockerClusterTemplate

File: controllers/topology/internal/mergepatch/managed_paths.go (hunk 1 above)
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/managed_paths.go:80]: Now we are respecting IngorePaths. We are already implicitly respecting ignore paths (they are hard coded, and we are considering only one of them - spec)

File: controllers/topology/internal/mergepatch/managed_paths.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+		paths := paths([]string{}, spec)
+		for _, p := range paths {
+			s = append(s, strings.Join(p, managedPathSeparator))
+		}
+
+		// Sort paths to get a predictable order (useful for readability and testing, not relevant at parse time).
+		sort.Strings(s)
+
+		// Concatenate all the paths in a single string.
+		// NOTE: add an extra space between fields for better readability (not relevant at parse time).
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/managed_paths.go:126]: I don't have strong opinions on the whitespaces, it just makes things more readable.
I'm also open to considering alternatives to the comma-separated list, do you have something in mind?

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 2 above)
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/mergepatch.go:164]: Nice catch
I have implemented a fix for managing changes on high-level objects. TL;DR;
- if a parent object is dropped from a template, we are deleting selectively what was set in a previous reconciliation (preserving nested user-defined values if any)
- if a parent object is explicitly set to null, the desired state aligns (dropping nested user-defined values if any)

File: api/v1beta1/common_types.go [hunk 4]
```diff
@@ -29,6 +29,19 @@ const (
 	// ClusterTopologyOwnedLabel is the label set on all the object which are managed as part of a ClusterTopology.
 	ClusterTopologyOwnedLabel = "topology.cluster.x-k8s.io/owned"
 
+	// ClusterTopologyManagedFieldsAnnotation is the annotation used to stores the list of paths managed
+	// by the topology controller; changes to those path will be considered autoritative.
+	// NOTE: Managed field depends by the last reconciliation of a managed object; this list can
```
Comment (sbueringer) [inline api/v1beta1/common_types.go:34]:
```suggestion
	// ClusterTopologyManagedFieldsAnnotation is the annotation used to store the list of paths managed
	// by the topology controller; changes to those path will be considered authoritative.
	// NOTE: Managed field depend on the last reconciliation of a managed object; this list can
```
nit

File: controllers/topology/internal/mergepatch/managed_paths.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply controllers/topology/internal/mergepatch/managed_paths.go:126]: One idea to optimize the footprint slightly would be to aggregate leaf nodes like this (with `{}`: (new lines are just here for illustration and shouldn't be part of the annotation)
```yaml
kubeadmConfigSpec.clusterConfiguration.apiServer.certSANs,
kubeadmConfigSpec.clusterConfiguration.imageRepository,
kubeadmConfigSpec.initConfiguration.nodeRegistration.criSocket,
kubeadmConfigSpec.initConfiguration.nodeRegistration.kubeletExtraArgs.{eviction-hard,cgroup-driver}
kubeadmConfigSpec.joinConfiguration.nodeRegistration.kubeletExtraArgs.eviction-hard,
machineTemplate.infrastructureRef.{apiVersion,kind,name,namespace}
machineTemplate.metadata.labels.{cluster.x-k8s.io/cluster-name,topology.cluster.x-k8s.io/owned},
replicas,
rolloutStrategy.rollingUpdate.maxSurge,
rolloutStrategy.type,
version
```

Of course this would also work on all layers:
```yaml
kubeadmConfigSpec.{clusterConfiguration.{apiServer.{certSANs},imageRepository,},initConfiguration{nodeRegistration.{criSocket,kubeletExtraArgs.{eviction-hard,cgroup-driver}}}
...
```

While the first option os still readable the second one gets quite obscure and it's probably not worth the risk  to introduce bugs.

[the remark this answers is not in the document]
Comment (sbueringer) [reply controllers/topology/internal/mergepatch/managed_paths.go:126]: P.S. I couldn't figure out a case where it fails, but can you think of a potential bug because dots in labels cannot be differentiated from the dots of nested fields?

File: controllers/topology/internal/mergepatch/mergepatch.go [hunk 5]
```diff
@@ -259,11 +277,24 @@ func enforcePath(authoritative, twoWay map[string]interface{}, path contract.Pat
...
+			}
 
-		// If the value is not a map (is a value or a list), return (not a full match).
-		if !ok {
-			return
+			// Otherwise, we continue processing the enforced path, thus deleting only what
+			// is explicitly enforced.
+			nestedSimpleMap = map[string]interface{}{path[1]: nil}
+		default:
+			nestedSimpleMap, _ = v.(map[string]interface{})
```
Comment (sbueringer) [inline controllers/topology/internal/mergepatch/mergepatch.go:297]: Shouldn't this throw a panic if `ok == false && v == nil` ?


```go
	v := map[string]interface{}{
		"abc": nil,
	}
	a := v["abc"].(map[string]interface{}) // panic
```

File: controllers/topology/internal/mergepatch/mergepatch.go [hunk 6]
```diff
@@ -259,11 +277,24 @@ func enforcePath(authoritative, twoWay map[string]interface{}, path contract.Pat
...
-		nestedSimpleMap, ok := authoritative[path[0]].(map[string]interface{})
+		var nestedSimpleMap map[string]interface{}
+		switch v, ok := authoritative[path[0]]; {
+		case ok && v == nil:
+			// If the nested map is nil, it means that the authoritative patch is trying to delete a parent object
+			// in the middle of the enforced path.
+
+			// If the parent object has been intentionally deleted (the corresponding parent object value in the modified object is null),
+			// then we should enforce the deletion of the parent object including everything below it.
+			if _, ok := modified[path[0]]; ok && v == nil {
```
Comment (sbueringer) [inline controllers/topology/internal/mergepatch/mergepatch.go:288]: v == nil is always true as we check it in l.282

File: controllers/topology/internal/mergepatch/mergepatch.go [hunk 7]
```diff
@@ -259,11 +277,24 @@ func enforcePath(authoritative, twoWay map[string]interface{}, path contract.Pat
...
+				twoWay[path[0]] = nil
+				return
+			}
 
-		// If the value is not a map (is a value or a list), return (not a full match).
-		if !ok {
-			return
+			// Otherwise, we continue processing the enforced path, thus deleting only what
+			// is explicitly enforced.
+			nestedSimpleMap = map[string]interface{}{path[1]: nil}
```
Comment (sbueringer) [inline controllers/topology/internal/mergepatch/mergepatch.go:295]: Shouldn't those maps only contain one path segment per layer?

File: controllers/topology/internal/mergepatch/mergepatch.go [hunk 8]
```diff
@@ -58,22 +58,33 @@ func NewHelper(original, modified client.Object, c client.Client, opts ...Helper
 		{"spec"},
```
Comment (sbueringer) [inline controllers/topology/internal/mergepatch/mergepatch.go:58]: Should we add a note that the managedField logic is only correct if the whole "spec" tree is part of `allowedPaths`?

I know we probably will never change this. It just feels not ideal having this dependency without at least a comment here that this dependency exists.

File: controllers/topology/internal/mergepatch/managed_paths_test.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+		obj  client.Object
+		want []contract.Path
+	}{
+		{
+			name: "Return no paths if the annotation does not exists",
+			obj: &unstructured.Unstructured{
+				Object: map[string]interface{}{},
+			},
+			want: []contract.Path{},
+		},
```
Comment (sbueringer) [inline controllers/topology/internal/mergepatch/managed_paths_test.go:42]: Maybe add a test case with an empty annotation and maybe just a bigger one something like
```bash
 kubeadmConfigSpec.clusterConfiguration.apiServer.certSANs,
        kubeadmConfigSpec.clusterConfiguration.imageRepository, kubeadmConfigSpec.initConfiguration.nodeRegistration.criSocket,
        kubeadmConfigSpec.initConfiguration.nodeRegistration.kubeletExtraArgs.eviction-hard,
        kubeadmConfigSpec.joinConfiguration.nodeRegistration.kubeletExtraArgs.cgroup-driver,
        kubeadmConfigSpec.joinConfiguration.nodeRegistration.kubeletExtraArgs.eviction-hard,
        machineTemplate.infrastructureRef.apiVersion, machineTemplate.infrastructureRef.kind,
        machineTemplate.infrastructureRef.name, machineTemplate.infrastructureRef.namespace,
        machineTemplate.metadata.labels.cluster.x-k8s.io/cluster-name, machineTemplate.metadata.labels.topology.cluster.x-k8s.io/owned,
        replicas, rolloutStrategy.rollingUpdate.maxSurge, rolloutStrategy.type, version
```

File: controllers/topology/internal/mergepatch/managed_paths.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/managed_paths.go:126]: we are including only field names in the label value, so there is no concern about undesired dots being added

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 6 above)
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/mergepatch.go:288]: fixed, this should be m (the value from modified, not v)

File: controllers/topology/internal/mergepatch/managed_paths.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply controllers/topology/internal/mergepatch/managed_paths.go:126]: No my point is more like, we are (currently) adding names of individual labels and those labels can contain dots. I was just wondering if that's an issue because we can't know if it's just a dot in a label or one of the nested field

E.g.:
```yaml
 kubeadmConfigSpec.clusterConfiguration.apiServer.certSANs,
        kubeadmConfigSpec.clusterConfiguration.imageRepository, kubeadmConfigSpec.initConfiguration.nodeRegistration.criSocket,
        kubeadmConfigSpec.initConfiguration.nodeRegistration.kubeletExtraArgs.eviction-hard,
        kubeadmConfigSpec.joinConfiguration.nodeRegistration.kubeletExtraArgs.cgroup-driver,
        kubeadmConfigSpec.joinConfiguration.nodeRegistration.kubeletExtraArgs.eviction-hard,
        machineTemplate.infrastructureRef.apiVersion, machineTemplate.infrastructureRef.kind,
        machineTemplate.infrastructureRef.name, machineTemplate.infrastructureRef.namespace,
        machineTemplate.metadata.labels.cluster.x-k8s.io/cluster-name, machineTemplate.metadata.labels.topology.cluster.x-k8s.io/owned,
        replicas, rolloutStrategy.rollingUpdate.maxSurge, rolloutStrategy.type, version
```

> machineTemplate.metadata.labels.topology.cluster.x-k8s.io/owned,

> machineTemplate.metadata.labels.
Those dots are all from the field

> topology.cluster.x-k8s.io/owned,
Those dots are just part of the label name.

I would have thought that this could lead to issues when keeping/dropping labels, but it doesn't seem to.

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 6 above)
Comment (sbueringer) [reply controllers/topology/internal/mergepatch/mergepatch.go:288]: Thought so

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 5 above)
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/mergepatch.go:297]: v being nil is handled by the first case in the switch statement; everything else should always be a map[string]interface{} given how Unmarshal works, but I have added a safeguard to avoid panics down the line.

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 7 above)
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/mergepatch.go:295]: we are adding only path[1] (with value nil)

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 8 above)
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/mergepatch.go:58]: done!

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 7 above)
Comment (sbueringer) [reply controllers/topology/internal/mergepatch/mergepatch.go:295]: Sorry misread as `path[1:]` :/

File: controllers/topology/internal/mergepatch/mergepatch.go (hunk 5 above)
Comment (sbueringer) [reply controllers/topology/internal/mergepatch/mergepatch.go:297]: My thought was that the first case only handles `ok && v==nil` but if `ok == false && v == nil` is impossible that's fine too :)

File: controllers/topology/internal/mergepatch/managed_paths_test.go (hunk 9 above)
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/managed_paths_test.go:42]: Added, it is kind of overkill for the first layer of unit test - considering we have 4 o 5 levels above - but I don't mind expanding a the existing deep nesting test

Comment (sbueringer) [reply controllers/topology/internal/mergepatch/managed_paths_test.go:42]: Ah sorry. Don't ask me how but I somehow missed your pre-existing KCP test.

**Comment by sbueringer** [comment]
lgtm pending the following from my side:
* https://github.com/kubernetes-sigs/cluster-api/pull/5812#discussion_r766799711
* https://github.com/kubernetes-sigs/cluster-api/pull/5812#discussion_r766826312 (I meant the other test)

File: controllers/topology/internal/mergepatch/managed_paths.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/managed_paths.go:126]: confirmed, this is not a problem.

[the remark this answers is not in the document]
Comment (sbueringer) [reply controllers/topology/internal/mergepatch/managed_paths.go:126]: I played around with it some more and was able to find the issue.

It turns it only did work because I dropped a label with dots from a path which was an authoritativePath anyway (spec.machineTemplate.metadata.labels). So it did only work because it was below spec.machineTemplate.metadata and not because of managed fields.

If I try to do this with some other path it fails:
https://github.com/kubernetes-sigs/cluster-api/pull/5846/commits/f6eee6f872d2177d1c5bba734cb0a0be24f87b11

[the remark this answers is not in the document]
Comment (sbueringer) [reply controllers/topology/internal/mergepatch/managed_paths.go:126]: I think we can only fix this by:
* using characters which are not valid character in labels as separator (dot is valid in labels, not sure about commas) or
* escaping labels (dots, maybe also commas)

**Comment by sbueringer** [comment]
lgtm pending squash + golangci-lint

File: controllers/topology/internal/mergepatch/managed_paths_test.go [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+				{"spec", "foo"},
+			},
+		},
+		{
+			name: "Handle deep nesting ",
+			obj: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"annotations": map[string]interface{}{
+							clusterv1.ClusterTopologyManagedFieldsAnnotation: "kubeadmConfigSpec.clusterConfiguration.imageRepository, " +
```
Comment (vincepri) [inline controllers/topology/internal/mergepatch/managed_paths_test.go:81]: Should we consider using some sort of encoding for this annotation? The paths can be easily compressed, and we could b64 encode the output

Comment (fabriziopandini) [reply controllers/topology/internal/mergepatch/managed_paths_test.go:81]: I kind of like the idea of this being human readable, so was considering something like 
`{ kubeadmConfigSpec { clusterConfiguration { imageRepository, version }, initConfiguration { ... } } }`
But this requires a little encoder/parser and I did not include in this first iteration, but I really want to sort it out before the actual release

File: api/v1beta1/common_types.go [hunk 11]
```diff
@@ -29,6 +29,19 @@ const (
...
+	// NOTE: Managed field depends on the last reconciliation of a managed object; this list can
+	// change during the lifecycle of an object, depending on how the corresponding template + patch/variable
+	// changes over time.
+	// NOTE: The topology controller is only concerned about managed paths in the spec; given that
+	// we are dropping spec. from the result to reduce verbosity of the generated annotation.
+	// NOTE: Managed paths are relevant only for unstructured objects where it is not possible
+	// to easily discover which fields have been set by templates + patches/variables at a given reconcile;
+	// instead, it is not necessary to store managed paths for typed objets (e.g. Cluster, MachineDeployments)
+	// given that the topology controller explicitly sets a well-known, immutable list of fields at every reconcile.
+	ClusterTopologyManagedFieldsAnnotation = "topology.cluster.x-k8s.io/managed-fields"
```
Comment (vincepri) [inline api/v1beta1/common_types.go:43]:
```suggestion
	ClusterTopologyManagedFieldsAnnotation = "topology.cluster.x-k8s.io/managed-field-paths"
```

Comment (fabriziopandini) [reply api/v1beta1/common_types.go]: Fixed!

**Review by vincepri — APPROVED** [review]
/approve
/lgtm
